snorkel/labeling/lf/nlp.py

```diff
@@ -0,0 +1,202 @@
+
+    Raises
+    ------
+    ValueError
+        Calling incorrectly defined preprocessors
+    """
+
+    _nlp_config: SpacyPreprocessorConfig
+
+    @classmethod
+    def _create_or_check_preprocessor(
+        cls,
+        text_field: str,
+        doc_field: str,
+        language: str,
+        disable: Optional[List[str]],
+        preprocessors: List[BasePreprocessor],
+        memoize: bool,
+    ) -> None:
+        # Create a SpacyPreprocessor if one has not yet been instantiated.
+        # Otherwise, check that configuration matches already instantiated one.
+        parameters = SpacyPreprocessorParameters(
+            text_field=text_field,
+            doc_field=doc_field,
+            language=language,
+            disable=disable,
+            preprocessors=preprocessors,
+            memoize=memoize,
+        )
+        if not hasattr(cls, "_nlp_config"):
+            nlp = SpacyPreprocessor(**parameters._asdict())
+            cls._nlp_config = SpacyPreprocessorConfig(nlp=nlp, parameters=parameters)
+        elif parameters != cls._nlp_config.parameters:
+            raise ValueError(
+                "NLPLabelingFunction already configured with different parameters: "
+                f"{cls._nlp_config.parameters}"
+            )
+
+    def __init__(
+        self,
+        name: str,
+        f: Callable[..., int],
+        resources: Optional[Mapping[str, Any]] = None,
+        preprocessors: Optional[List[BasePreprocessor]] = None,
+        fault_tolerant: bool = False,
+        text_field: str = "text",
+        doc_field: str = "doc",
+        language: str = EN_CORE_WEB_SM,
+        disable: Optional[List[str]] = None,
+        memoize: bool = True,
+    ) -> None:
+        self._create_or_check_preprocessor(
+            text_field, doc_field, language, disable, preprocessors or [], memoize
+        )
+        super().__init__(
+            name,
+            f,
+            resources=resources,
+            preprocessors=[self._nlp_config.nlp],
+            fault_tolerant=fault_tolerant,
+        )
+
+
+class nlp_labeling_function:
+    """Decorator to define an NLPLabelingFunction object from a function.
+
+    Parameters
+    ----------
+    See `NLPLabelingFunction`.
+
+
+    Examples
+    --------
+    ```
+    @nlp_labeling_function()
+    def has_person_mention(x: DataPoint) -> int:
+        person_ents = [ent for ent in x.doc.ents if ent.label_ == "PERSON"]
+        return 1 if len(person_ents) > 0 else 0
+    print(f)  # "NLPLabelingFunction has_person_mention"
+
+    x = SimpleNamespace(text="The movie was good.")
+    has_person_mention(x)  # 0
+    ```
+    """
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        resources: Optional[Mapping[str, Any]] = None,
+        preprocessors: Optional[List[BasePreprocessor]] = None,
+        fault_tolerant: bool = False,
+        text_field: str = "text",
+        doc_field: str = "doc",
+        language: str = EN_CORE_WEB_SM,
```

Comment:
can we add a small comment about what this refers to?

Reply:
included in `NLPLabelingFunction`, which is referenced by this decorator. is that enough?

Reply:
sg!